### PR TITLE
Make a post on the website stating Armory's position on BU

### PR DESCRIPTION
**Pull request by achow101**
This pull request is for a post for Armory's website that states Armory's position on BU and how to taint coins should a hard fork occur. The instructions are adapted from https://bitcointalk.org/index.php?topic=1837143.msg18289746#msg18289746

@goatpig @droark Please review this and let me know if it should be revised.

The page on the website will look like this:

image

### CODE REVIEW

**Comment by goatpig** [comment]
I don't know that controversial is a good word to describe the current state of affairs. Basically it should say Armory does not care about forks because wallets do not care about forks, and aren't involved in politics to begin with.

But Armory opposes and denounces all fork splits that come with an attack on the original chain. We won't stand with that stuff and if BU commits that crime, they'll the associated treatment.

**Comment by goatpig** [comment]
If you want a statement about endorsing Core, you can write a snippet saying something like "as an Armory dev, I personally endorse Core" and add a "signed by goatpig, achow101 and droark (I believe Doug is pro Core).

**Comment by achow101** [comment]
I'm not sure about the "attack on the original chain" part. I don't think BU is going to do that. Basically what I wanted to say in the first sentence was that if BU forked and then deployed FlexTrans, we wouldn't implement FlexTrans for those using BU.

**Comment by goatpig** [comment]
I will personally not implement anything BU comes up with unless it is necessary for people to move their mirrored coins. If say FlexTrans was a hardfork and compulsory, I'd have to consider integrating towards it to allow users to moving their coins, if the circumstances are overwhelmingly in favor (i.e. there's no other option for Armory users for a long time).

On other hand if users can do without the support, or FlexTrans is an opt-in feature, I would ignore it as a whole. However if someone forks the repo, implements support for it and PRs that back to me, I will at least review the change set. If it's good and segregated, I will merge it.

This obviously implies BU fork that survives at least for a year. Not gonna merge code for ephemeral splits.

As for the BU attack, my stance is clearly "if they fork and attack". The threats are bad but don't constitute an aggression on their own.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+---
+layout: post
+title: Armory and Hard Forks
+category: news
+date: 2017-03-26 23:30:00 -0500
+---
+
+Armory does not support any controversial hard forks such as Bitcoin Unlimited and will continue to support Segregated Witness and Bitcoin Core.
```
Comment (achow101) [inline _posts/2017-3-26-armory-and-hard-forks.md]: How about changing this line to 

> The Armory developers (goatpig, droark, and achow101) oppose hard forks that may attack the original chain and support Segregated Witness and Bitcoin Core. Should a long-lasting hard fork that does not attack the original chain exist, we would implement functionality required to allow Armory users to transact on that chain only if absolutely necessary (e.g. the transaction format for that chain is changed).

The second sentence would require a bit of rewording.

**Comment by achow101** [comment]
I'll squash if this change looks good to everyone.

**Comment by droark** [comment]
Hello. Here's my take on things, along with a few suggested revisions.

\*takes off developer cap\* Personally, I'm 100% pro-Core. I've met some of the BU people, and am familiar with some of the tactics they're using behind the scenes. I'd rather quit working on Bitcoin altogether than lend any sort of legitimacy to what they've been doing. Core devs may be a bit tone-deaf at times - the HK agreement was, IMO, a terrible idea in retrospect - but they have the chops to actually keep Bitcoin going and improving without trashing the things that make Bitcoin interesting. (That and not trashing the code because they have no clue how to keep their client running.) I'll stop here.

\*puts cap back on\* Now then, I'd prefer that any official statements minimize politics whenever possible; it's best for everybody in the long run. The statement, as-is, seems reasonable in that regard. It is important to make clear that intentionally malicious forks are unwelcome in the broader community. (To me, malicious = Direct attempts to force people to use one chain over another.) Hard forks that naturally gain support may be open to some sort of migration path, or even full support.

Here are my suggested revisions. They're primarily grammar-related. If desired, I'll move the comments over to an actual review.

> The Armory developers (goatpig, droark, and achow101) oppose hard forks that may attack the original chain and support Segregated Witness and Bitcoin Core.

The Armory developers (goatpig, droark, and achow101) oppose hard forks that may attack the original chain. The Armory developers support Segregated Witness and Bitcoin Core.

> (End of first paragraph)

I'm wondering if it's worth stating that, as Armory is FOSS, people are welcome to fork Armory and support non-Core chains. I'm not going to work on a BU fork if I can help it. I'm also fine with others running with the ball if they see fit and can maintain their fork.

> (End of third paragraph)

I'd add another paragraph before the setup section. My understanding is that, while the suggested method *should* work, there's no guarantee. The assumption is that BU would have enough hashpower to handle lower fees than Core. I think it's a reasonable assumption. It's still an assumption. :) Depending on this & that, it could also be a temporary solution that works only while Core's rebounding from the hash drop or other problems.

Anyway, the upshot is that we should add a disclaimer, and ask people to keep an eye on the website, forums, etc. for up-to-date info in case a fork does occur. Standard "no warranty implied, proceed at your own peril, etc." kinda stuff.

> The DB builds off of on disk chain data and does not enforce consensus rules.

The database builds off of on-disk chain data and does not enforce consensus rules.

> Same goes with the Armory DB, as it is intimately tied to the underlying block data.

Same goes with the Armory database, as it is intimately tied to the underlying block data.

> However, you do not need to rescan the BU chain to get the copy of Core DB to work off of it.

I believe this is true only if the user hasn't received coins post-fork? This statement may need to be clarified.

Also, I'd point out that, post-fork, it won't be safe to make a simple copy of the chain. If a fork happens, people will need to either do a clean download if they want to guarantee a functional blockchain. A pre-fork copy *should* be fine, although again, there are no strict guarantees.

> You can swap between the 2 by pointing Armory to the relevant pair of block data folder (--satoshi-datadir) and database folder (--dbdir).

You can swap between the two by pointing Armory to the relevant pair of block data folder and database folder (--satoshi-datadir and --dbdir command line options respectively).

> Avoiding transaction replay

I'd add a short paragraph explaining Tx replay attacks. Nothing fancy, just something for the less technically inclined users.

> At this point none of your utxos on the one chain are valid on the other, which guarantees your txs aren't replayable.

At this point, none of your UTXOs on one chain are valid on the other chain. This guarantees that your transactions are immune to replay attacks.

> Create a transaction sending coins back to yourself, with a fee low enough such that it will quickly mine on BU but has no real chance of mining on Core.

Create a transaction sending coins back to yourself, with a fee low enough such that it will quickly mine on BU and will probably not be mined quickly on Core.

> Once the tx is mined on BU, RBF the underlying output on the Core chain with a nice fee. This tx can only mine on Core now since it is invalid on BU (to which it appears as a double spend).

Once the Tx is mined on BU, RBF the underlying output on the Core chain with a sufficiently high fee. This Tx can only be mined on Core now since it is invalid on BU (to which the Tx appears as a double spend).

> Once your RBF gets a good confirmation cushion on the Core side, you have a utxo exlusive to the BU chain and one exclusive the Core chain. You are now safe to taint the rest of your stash on each chain, keeping in mind that untainted coins remain valid on both chains.

Once your RBF has enough confirmations on the Core side - six should be sufficient - you have a UTXO exclusive to the BU chain and one exclusive the Core chain. You are now safe to taint the rest of your coins on each chain, keeping in mind that untainted coins remain valid on both chains. Including untainted coins with tainted coins in another transaction is sufficient for tainting the untainted coins.

> There is essentially no risk to tainting your coins on your own, at worst you are just sending your coins back to yourself on both chains.

There is essentially no risk to tainting your coins on your own. At worst, you are just sending your coins back to yourself on both chains and losing a small amount of coins in fees. To be safe, we recommend starting with a small transaction that can be tainted with minimal risk, and then tainting the remaining coins all at once so as to minimize fee loss.

Misc: Personally, I like to spell out short numbers ("one" instead of "1"), capitalize some abbreviations ("Tx" and "UTXO"), and put a comma after acronyms ("i.e.,"). It's your call, though, as long as it's consistent. :)

Thanks.

**Comment by achow101** [comment]
I'll go through your comments and revise.

Pretty much everything after the third paragraph is copy-paste directly from @goatpig's post, so blame him for the errors :p

**Comment by achow101** [comment]
Revised

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
+---
+layout: post
+title: Armory and Hard Forks
+category: news
+date: 2017-03-26 23:30:00 -0500
+---
+
+The Armory developers (goatpig, droark, and achow101) oppose hard forks that may attack the original chain. The Armory developers support Segregated Witness and Bitcoin Core. Should a long-lasting hard fork that does not attack the original chain exist, we would consider implementing functionality required to allow Armory users to transact on that chain only if absolutely necessary (e.g. the transaction format for that chain is changed). Depending on how drastic the changes are, we may implement the change fully, implement a migration tool to another wallet that supports the change, or simply not support the fork at all.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: Almost there. I'd reword things slightly. (Missed this the first time. Sorry!) Opening "cold" is a bit strange. I've tossed in a quick intro and am going from there.

"Recently, there has been much debate in the Bitcoin community regarding Segregated Witness, soft forks vs. hard forks, and the usage of alternative miners and nodes. The Armory developers (goatpig, droark, and achow101) fully support Segregated Witness in its current soft fork state, and also support Bitcoin Core. The Armory developers also oppose hard forks that may attack the original chain."

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+Although the Armory developers do not support such hard forks, the Armory wallet software does not perform any consensus checks as it relies on its connection to a local Bitcoin node. This node is typically Bitcoin Core, but anything based upon that will work as well; this includes Bitcoin Unlimited. Thus should a hard fork occur, so long as no changes are made to the transaction or block format, Armory will be compatible with the hard fork and will allow users to continue to transact on the forked network.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "This node is typically Bitcoin Core, but anything based upon that will work as well; this includes Bitcoin Unlimited. Thus should a hard fork occur, so long as no changes are made to the transaction or block format, Armory will be compatible with the hard fork and will allow users to continue to transact on the forked network."

This node is typically Bitcoin Core, but any node based upon Core, including Bitcoin Unlimited, will also work so long as no changes are made to the transaction or block formats. So long as both formats do not change, Armory should be compatible with the hard fork and all users to continue to transact on the forked network.

(**NB**: I say "should" in order to be safe. If an Unlimited-style fork occurs, I'm pretty sure Armory will be fine. It's also theoretically possible that another hard fork down the road could leave the formats alone and still be incompatible somehow.)

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+Because transactions for the forked network and for the original will primarily be valid on both networks, there is the risk for replay attacks to occur and for the loss of coins due to transaction replay. The following will help users avoid transaction replay should Bitcoin Unlimited or another similar block size increase hard fork proposal be activated. However this procedure is not necessarily guaranteed to work and assumes that the forked chain will have significantly lower transaction fees than the original chain.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "However this procedure is not necessarily guaranteed to work and assumes that the forked chain will have significantly lower transaction fees than the original chain."

However, this procedure....

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+Should a chain fork occur, more up-to-date information about what Armory users can do will be posted on the [Armory forums](https://bitcointalk.org/index.php?board=97.0) and on this website.
+
+### The setup
+
+The Armory database engine builds off of on-disk blockchain data and does not enforce consensus rules. In other words, any client that uses the same block data format and magic word as Bitcoin Core is compatible with Armory. From your perspective, running Armory against the BU chain will show you your BU balance, and vice versa. 
+
+In practice, this means that to be able to interchange seamlessly between your Core and BU transaction history, you need 2 copies of the blockchain data; one for the Core chain and one for the BU chain. Same goes with the Armory database, as it is intimately tied to the underlying block data. However, you do not need to rescan the BU chain to get the copy of pre-fork Core database to work off of it. To be absolutely safe, you should use rebuild and rescan the databases for both chains after the fork, and even fully redownload the blockchain to ensure that everything is correct.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "Same goes with the Armory database, as it is intimately tied to the underlying block data. However, you do not need to rescan the BU chain to get the copy of pre-fork Core database to work off of it. To be absolutely safe, you should use rebuild and rescan the databases for both chains after the fork, and even fully redownload the blockchain to ensure that everything is correct."

The same goes...block data. You do not need to rescan...pre-fork Core database to work off of it. However, to be absolutely safe, you should use Armory's "Rebuild and Rescan" feature to update both databases after the fork. In addition, it may be necessary to fully re-download the blockchain if the copy is made post-fork.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+To summarize, you need:
+
+- 1 copy of the blockchain to run exclusively against the Core client, with the associated Armory database directory.
+- 1 copy of the blockchain to run exclusively against the BU client, with associated Armory database directory.
+
+You can swap between the two by pointing Armory to the relevant pair of block data folder and database folder by using the `--satoshi-datadir` and `--dbdir` command line options respectively.
+
+### Avoiding Transaction Replay
+
+A transaction replay attack is where one transaction can end up confirmed in both blockchains thus resulting in coins on one chain being spent when you did not want those to be spent. This can happen either accidentally or intentionally where someone broadcasts your transaction to both networks.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "A transaction replay attack is where one transaction can end up confirmed in both blockchains thus resulting in coins on one chain being spent when you did not want those to be spent. This can happen either accidentally or intentionally where someone broadcasts your transaction to both networks."

A transaction replay...both blockchains, thus resulting.... This can happen accidentally or intentionally due to someone broadcasting your transaction on both networks.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+In order to avoid transaction replay, you must "taint" your coins; you want to get a transaction on one chain that is not valid on the other. Once you have that, you can mix all of your remaining coins with the "tainted" output to the same effect. At this point none of your UTXOs on one chain are valid on the other, which guarantees that your transactions are immune to transaction replay attacks.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "UTXOs" --> [unspent transaction outputs (UTXOs)](https://bitcoin.org/en/glossary/unspent-transaction-output). Only needs to be spelled out once.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+Tainting on BU is actually fairly easy. They want to increase the block size limit and reduce tx fees. At the same time, the departure of hash power from the main chain will result in slower Core blocks and thus less capacity and higher fees.
+
+Hence, the simplest way to taint on BU is this:
+
+- Create a transaction sending coins back to yourself, with a fee low enough such that it will quickly mine on BU and will probably not be mined quickly on Core.
+
+- Once the transaction is mined on BU, RBF the underlying output on the Core chain with a sufficiently high fee. This transaction can only be mined on Core now since it is invalid on BU (to which the transaction appears as a double spend).
+
+- Once your RBF transaction receives enough confirmations on the Core side -- six should be sufficient, you will have a UTXO exlusive to the BU chain and a UTXO exclusive the Core chain. You are now safe to taint the rest of your coins on each chain, keeping in mind that untainted coins remain valid on both chains. Including untainted coins in a transaction with tainted coins will be sufficient for tainting the untainted coins.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "Once your RBF transaction receives enough confirmations on the Core side -- six should be sufficient, you will have a UTXO exlusive"

Once your RBF transaction receives enough confirmations on the Core side -- six should be sufficient **--** you will have a UTXO **exclusive**....

Comment (goatpig) [inline _posts/2017-3-26-armory-and-hard-forks.md]: In the OP, the context was accounting for a miner attack, so the delays advertised were very large (1-2 days). Is there room for that caveat here or do we not care?

**Comment by droark** [comment]
Updated with a few more comments. I should be happy once they're folded in.

File: _posts/2017-3-26-armory-and-hard-forks.md (hunk 8 above)
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: Good point. I would point out that, in a nightmare scenario, it could take days to confirm transactions. Is this okay to add to the end of the paragraph?

"Please note that, in a worst case scenario, the miner hashrate drop could mean days pass before a block is added to a chain. While unlikely, this means weeks, or even months, could pass before coins can be tainted safely."

**Comment by goatpig** [comment]
ACK so far. I'll review the final document once it congeals.

**Comment by achow101** [comment]
Revised with @droark's comments.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+Although the Armory developers do not support such hard forks, the Armory wallet software does not perform any consensus checks as it relies on its connection to a local Bitcoin node. This node is typically Bitcoin Core, but anything based upon Core, including Bitcoin Unlimited, will work so long as no changes are made to the transaction or block formats. So long as both formats do not change, Armory should be compatible with the hard fork and will allow users to continue to transact on the forked network.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "Although the Armory developers do not support such hard forks,"

This ("such hard forks") needs to be revised, although I'm not totally sure of the wording. I'm okay with hard forks that have strong consensus and don't actively attack other chains. I assume @goatpig is too? Perhaps "do not support **malicious** hard forks" is reasonable?

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+### The setup
+
+The Armory database engine builds off of on-disk blockchain data and does not enforce consensus rules. In other words, any client that uses the same block data format and magic word as Bitcoin Core is compatible with Armory. From your perspective, running Armory against the BU chain will show you your BU balance, and vice versa. 
+
+In practice, this means that to be able to interchange seamlessly between your Core and BU transaction history, you need 2 copies of the blockchain data; one for the Core chain and one for the BU chain. The same goes with the Armory database, as it is intimately tied to the underlying block data. You do not need to rescan the BU chain to get the copy of pre-fork Core database to work off of it. However, to be absolutely safe, you should use Armory's *Rebuild and Rescan Databases* feature to update both databases after the fork. In addition, it may be necessary to fully re-download the blockchain if the blockchain and database copies are made post-fork.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: "You do not need to rescan the BU chain to get the copy of pre-fork Core database to work off of it."

You do not need to rescan the **pre-fork** BU chain to get **a** copy of **the** pre-fork Core database to work off of **the pre-fork BU chain**.

File: _posts/2017-3-26-armory-and-hard-forks.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+Tainting on BU is actually fairly easy. They want to increase the block size limit and reduce tx fees. At the same time, the departure of hash power from the main chain will result in slower Core blocks and thus less capacity and higher fees.
+
+Hence, the simplest way to taint on BU is this:
+
+- Create a transaction sending coins back to yourself, with a fee low enough such that it will quickly mine on BU and will probably not be mined quickly on Core.
+
+- Once the transaction is mined on BU, RBF the underlying output on the Core chain with a sufficiently high fee. This transaction can only be mined on Core now since it is invalid on BU (to which the transaction appears as a double spend).
+
+- Once your RBF transaction receives enough confirmations on the Core side -- six should be sufficient -- you will have a UTXO exclusive to the BU chain and a UTXO exclusive the Core chain. You are now safe to taint the rest of your coins on each chain, keeping in mind that untainted coins remain valid on both chains. Including untainted coins in a transaction with tainted coins will be sufficient for tainting the untainted coins.
```
Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: No mention of the possibility of long waits for confirmations? I have no strong opinions either way, although I'd prefer it since non-technical people will inevitably flood the forums should things get that bad.

File: _posts/2017-3-26-armory-and-hard-forks.md (hunk 9 above)
Comment (achow101) [inline _posts/2017-3-26-armory-and-hard-forks.md]: How about

> Although the Armory developers do not support malicious hard forks and may not support a future non-malicious hard fork, ...

Basically a statement that although some hypothetical hard fork in the future is not supported, it won't necessarily cause Armory to fail.

Comment (droark) [inline _posts/2017-3-26-armory-and-hard-forks.md]: Sounds good.

**Comment by achow101** [comment]
revised again

**Comment by droark** [comment]
ACK

Some grammar nits and such remain. I just don't want to bikeshed this message. The gist of it looks good to me.

**Comment by achow101** [comment]
Doing a last minute grammar check right now.

**Comment by achow101** [comment]
ready for merge now
